### PR TITLE
Reserve Nodeport Ranges For Dynamic And Static Port Allocation

### LINKED ISSUE #114129: Reserved an statically port range for nodeport
### What would you like to be added?

just like [KEP-3070](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/3070-reserved-service-ip-range) did, I want to reserve some port for nodeport.  I can try to submit an KEP if it sound good for sig/network.

### Why is this needed?

In our cluster, some applications are NOT deployed in Kubernetes. But they need to connect to `service` in Kubernetes too.

For example, some applications need to connect to the `minio` service but they have no idea what cluster-ip it is. In our situation, those applications just connect to a specific port (30009) of the Kubernetes master node.

This situation works most time, but in something 30009 is allocated by something else. And minio service will be failed to applied.

### PR DESCRIPTION
**Pull request by xuzhenglun**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
quote from https://github.com/kubernetes/kubernetes/issues/114129:
> In our cluster, some applications are NOT deployed in Kubernetes. But they need to connect to service in Kubernetes too.
> For example, some applications need to connect to the minio service but they have no idea what cluster-ip it is. In our situation, those applications just connect to a specific port (30009) of the Kubernetes master node.
> This situation works most time, but in something 30009 is allocated by something else. And minio service will be failed to applied.

This PR implements [KEP 3682](https://github.com/kubernetes/enhancements/pull/3682):

- [X] subdivide node port range and allocate nodeport with offset
- [x] add 4 more metrics to improve observability of this feature

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/114129

#### Does this PR introduce a user-facing change?
```release-note
New feature gate, ServiceNodePortStaticSubrange, to enable the new strategy in the NodePort Service port allocators, so the node port range is subdivided and dynamic allocated NodePort port for Services are allocated preferentially from the upper range.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/3682
```

### CODE REVIEW

File: pkg/registry/core/service/allocator/interfaces.go [hunk 1]
```diff
@@ -44,6 +44,6 @@ type Snapshottable interface {
 	Restore(string, []byte) error
 }
 
-type AllocatorFactory func(max int, rangeSpec string) (Interface, error)
+type AllocatorFactory func(max int, rangeSpec string, offset int) (Interface, error)
```
Comment (aojea) [inline pkg/registry/core/service/allocator/interfaces.go:47]: why not using `AllocatorWithOffsetFactory`?

File: pkg/registry/core/service/storage/storage_test.go [hunk 2]
```diff
@@ -39,6 +39,8 @@ import (
 	genericregistrytest "k8s.io/apiserver/pkg/registry/generic/testing"
 	"k8s.io/apiserver/pkg/registry/rest"
 	etcd3testing "k8s.io/apiserver/pkg/storage/etcd3/testing"
+	netutils "k8s.io/utils/net"
```
Comment (aojea) [inline pkg/registry/core/service/storage/storage_test.go:42]: this change doesn't seem related to this PR

File: pkg/registry/core/service/allocator/interfaces.go (hunk 1 above)
Comment (aojea) [reply pkg/registry/core/service/allocator/interfaces.go:47]: do you really need to modify this?

Comment (xuzhenglun) [reply pkg/registry/core/service/allocator/interfaces.go]: well, I didn't notice it. `AllocatorWithOffsetFactory` is better, I correct it now.

File: pkg/features/kube_features.go [hunk 3]
```diff
@@ -908,6 +908,13 @@ const (
 	// instead of changing each file on the volumes recursively.
 	// Initial implementation focused on ReadWriteOncePod volumes.
 	SELinuxMountReadWriteOncePod featuregate.Feature = "SELinuxMountReadWriteOncePod"
+
+	// owner: @xuzhenglun
+	// kep: http://kep.k8s.io/3682
+	// alpha: v1.27
+	//
+	// Subdivide the NodePort range for dynamic and static port allocation.
+	ServiceNodePortStaticSubrange featuregate.Feature = "ServiceNodePortStaticSubrange"
```
[the remark this answers is not in the document]
Comment (xuzhenglun) [reply pkg/features/kube_features.go]: Well, it's true that the current feature gates are not completely sorted, so I'll insert them in the current order, and is that ok for you?

[the remark this answers is not in the document]
Comment (aojea) [reply pkg/features/kube_features.go]: yes, thanks

File: pkg/registry/core/service/portallocator/metrics.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	// nodePortAllocated indicates the amount of ports allocated by NodePort Service.
+	nodePortAllocated = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "allocated_ports",
+			Help:           "Gauge measuring the number of allocated NodePorts for Services",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"cidr"},
```
Comment (aojea) [inline pkg/registry/core/service/portallocator/metrics.go:41]: you don't need  a vector in all this metrics, the NodePorts are global, there is no range per ip family

File: pkg/registry/core/service/portallocator/allocator.go [hunk 5]
```diff
@@ -123,33 +128,57 @@ func (r *PortAllocator) Used() int {
 // or has already been reserved.  ErrFull will be returned if there
 // are no ports left.
 func (r *PortAllocator) Allocate(port int) error {
+	label := r.portRange.String()
```
Comment (aojea) [inline pkg/registry/core/service/portallocator/allocator.go:131]: commented on the metrics

File: pkg/registry/core/service/portallocator/metrics.go (hunk 4 above)
Comment (xuzhenglun) [reply pkg/registry/core/service/portallocator/metrics.go]: you are right, It's done

**Comment by xuzhenglun** [comment]
@aojea 

KEP is merged now, please take a look at this PR when you are convenient.

---

simply run an e2e test and here is the difference between enabling `ServiceNodePortStaticSubrange` or not.

image

File: pkg/registry/core/service/portallocator/allocator.go [hunk 6]
```diff
@@ -213,3 +256,36 @@ func (r *PortAllocator) contains(port int) (bool, int) {
...
+func calculateRangeOffset(pr net.PortRange) int {
+	// default values for min(max($min, rangeSize/$step), $max)
+	const (
+		min  = 16
+		max  = 128
+		step = 32
+	)
+
+	rangeSize := pr.Size
+	if rangeSize < min {
```
Comment (aojea) [inline pkg/registry/core/service/portallocator/allocator.go:279]: is not the same case because subnets size are different, but better use <= to avoid possible problems like this

https://github.com/kubernetes/kubernetes/pull/115322

Comment (xuzhenglun) [reply pkg/registry/core/service/portallocator/allocator.go]: It's done.

**Comment by aojea** [comment]
/lgtm
/approve

/assign @thockin 

Thanks

**Comment by xuzhenglun** [comment]
ping @thockin
